packages/@react-spectrum/tag/src/Tag.tsx

```diff
@@ -10,82 +10,105 @@
-    role
-  });
-  let {role: buttonRole, ...otherButtonProps} = clearButtonProps;
-  let icon = props.icon || (isInvalid && <Alert />);
+    isRemovable,
+    isDisabled,
+    item,
+    onRemove,
+    tagRef,
+    labelRef
+  }, state);
 
   return (
-    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
-      <div
-        {...styleProps}
-        {...mergeProps(tagProps, hoverProps)}
-        className={classNames(
-          styles,
-          'spectrum-Tags-item',
-          {
-            'is-disabled': disabled,
-            // 'is-selected': isSelected,
-            'spectrum-Tags-item--removable': removable,
-            'is-invalid': isInvalid,
-            'is-hovered': isHovered
-          },
-          styleProps.className
-        )}>
-        {icon && React.cloneElement(icon, {size: 'S', UNSAFE_className: classNames(styles, 'spectrum-Tags-itemIcon')})}
-        <span
-          {...labelProps}
-          className={classNames(styles, 'spectrum-Tags-itemLabel')}>
-          {props.children}
-        </span>
-        {removable &&
-          <span role={buttonRole}>
-            <ClearButton
-              tabIndex={tagProps.tabIndex}
-              focusClassName={classNames(styles, 'is-focused')}
-              UNSAFE_className={classNames(styles, 'spectrum-Tags-itemClearButton')}
-              {...otherButtonProps} />
-          </span>
+    <div
+      {...mergeProps(tagProps, hoverProps, focusProps, otherProps)}
+      role={tagProps.role}
+      className={classNames(
+        styles,
+        'spectrum-Tags-item',
+        {
+          'is-disabled': isDisabled,
+          'focus-ring': isFocusVisible,
+          'is-focused': isFocusVisible,
+          'not-removable': !isRemovable,
+          'is-hovered': isHovered
         }
-      </div>
-    </FocusRing>
+      )}>
+      <SlotProvider
+        slots={{
+          tagLabel: {UNSAFE_className: classNames(styles, 'spectrum-Tag-label')}
+        }}>
+        <TagLabel {...mergeProps(props, labelProps)} isRemovable={isRemovable} />
+        {isRemovable && <TagRemoveButton item={item} {...clearButtonProps} UNSAFE_className={classNames(styles, 'spectrum-Tag-action')} />}
+      </SlotProvider>
+    </div>
+  );
+}
+
+function TagLabel(props) {
+  let {role, children, isRemovable} = props;
+  props = useSlotProps(props, 'tagLabel');
+  let {styleProps} = useStyleProps(props);
+  let labelRef = useRef();
+
+  console.log(isRemovable);
+
+  return (
+    <span
+      {...styleProps}
+      role="rowheader"
+      ref={labelRef}
+      className={classNames(styles, 'spectrum-Tags-labelItem')}>
+      <SlotProvider
+        slots={{
+          icon: {UNSAFE_className: classNames(styles, 'spectrum-Tag-icon'), size: 'XS'},
+          text: {UNSAFE_className: classNames(styles, 'spectrum-Tag-content', {'tags-removable': isRemovable})}
+        }}>
+        {typeof children === 'string' ? <span role={role}><Text>{children}</Text></span> : children}
```

Comment:
The `role="rowheader"` should not contain another element with `role="gridcell"`, they are mutually exclusive. The element with `role="rowheader"` should contain the icon svg element and the label span element, and the label span element should have no role.

Reply:
I'm also able to reproduce the crash reported by @LFDanLu here: https://github.com/adobe/react-spectrum/pull/1922#pullrequestreview-823044078, which makes it impossible to test keyboard interaction.